### PR TITLE
Add temporality to MetricDescriptor

### PR DESCRIPTION
**Pull request by MrAlias**
Updates the `MetricDescriptor` to specify the temporal quality of the data being transported. Consequentially, `type` is now decoupled from this introduced concept.

This is a partial replacement of #137 
Relates to #125

### CODE REVIEW

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 1]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
   }
   Type type = 4;
 
+  // Temporality is the temporal quality values of a metric have. It
+  // describes how those values relate to the time interval over which they
+  // are reported.
+  enum Temporality {
+    // TEMPORALITY_INVALID is the default Temporality, it MUST not be
+    // used.
+    TEMPORALITY_INVALID = 0;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:159]: UNKNOWN (or just INVALID temporality is included in the generated name, this will be TEMPORALITY_TEMPORALITY_INVALID) is usually the recommended name for this by the proto3 conventions https://developers.google.com/protocol-buffers/docs/proto3#enum

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 2]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+  // Monotonic is a refinement of the values a metric has. It defines the
+  // relationship values of successively reported metrics have
+  // (non-increasing, non-decreasing, or unknown). This is a refinement of
+  // the metric values that can be useful for a receiver in understanding
+  // how to deal with discontinuities in the data (i.e.  calculating
+  // derivates of the data without introducing artifacts from a reset).
+  enum Monotonic {
+    // MONOTONIC_UNSPECIFIED is the default, and means the monotonic nature
+    // of the metric values is unknown.
+    MONOTONIC_UNSPECIFIED = 0;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:238]: Same just "UNSPECIFIED"

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 3]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+
+  // monotonic describes the Monotonic refinement of values this metric has.
+  Monotonic monotonic = 6;
+
+  // Domain is a refinement of the values a metric has. It describes the set
+  // of numbers metric values belong to, if any.
+  enum Domain {
+    // DOMAIN_UNSPECIFIED is the default, and means the metric values do
+    // not belong to any particular domain other than the Type itself.
+    DOMAIN_UNSPECIFIED = 0;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:253]: Same just "UNSPECIFIED"

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 4]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+  // are reported.
+  enum Temporality {
+    // TEMPORALITY_INVALID is the default Temporality, it MUST not be
+    // used.
+    TEMPORALITY_INVALID = 0;
+
+    // INSTANTANEOUS is a metric whose values are measured at a particular
+    // instant. The values are not aggregated over any time interval and are
+    // unique per timestamp.
+    INSTANTANEOUS = 1;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:164]: Do we want to specify that a start time is not expected for this?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 5]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+    // made over a time interval. Successive metrics contain aggregation of
+    // values from continuous and non-overlapping intervals.
+    //
+    // The values for a DELTA metric are based only on the time interval
+    // associated with one measurement cycle. There is no dependency on
+    // previous measurements like is the case for CUMULATIVE metrics.
+    //
+    // For example, consider a system measuring the number of requests that
+    // it receives every second and reports the sum of these requests as a
+    // DELTA metric:
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:176]:
```
    // For example, consider a system measuring the number of requests that
    // it receives and reports the sum of these requests every second as a
    // DELTA metric:
```

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 6]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+    // reported measurement. This means that current values of a CUMULATIVE
+    // metric depend on all previous measurements since the start time.
+    // Because of this, the sender is required to retain this state in some
+    // form. If this state is lost or invalidated, the CUMULATIVE metric
+    // values MUST be reset and a new fixed start time following the last
+    // reported measurement time sent MUST be used.
+    //
+    // For example, consider a system measuring the number of requests that
+    // it receives every second and reports the sum of these requests as a
+    // CUMULATIVE metric:
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:203]: Same you don't measure received requests every second, you measure requests received and export every second.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 7]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+    //      t_0+1 with a value of 1.
+    CUMULATIVE = 3;
+  }
+
+  // temporality is the Temporality of values this metric has.
+  Temporality temporality = 5;
+
+  // Monotonic is a refinement of the values a metric has. It defines the
+  // relationship values of successively reported metrics have
+  // (non-increasing, non-decreasing, or unknown). This is a refinement of
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:231]: Consider to update the comment, no longer support for non-increasing.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 8]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+  // the metric values that can be useful for a receiver in understanding
+  // how to deal with discontinuities in the data (i.e.  calculating
+  // derivates of the data without introducing artifacts from a reset).
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:234]: Also it is used to determine if calculating a rate makes sense or not (for example if a metric is not monotonic increasing then the rate, if reported t0->v0 and t1->v1, (v1 - v0) / (t1 - t0) may not be accurate because the value may jumped and come back to v1.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 9]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+  enum Monotonic {
+    // MONOTONIC_UNSPECIFIED is the default, and means the monotonic nature
+    // of the metric values is unknown.
+    MONOTONIC_UNSPECIFIED = 0;
+
+    // NONDECREASING means all the successive metric values increase or
+    // remain constant.
+    NONDECREASING = 1;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:242]: Curios why using negative forms?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 10]
```diff
@@ -172,9 +172,118 @@ message MetricDescriptor {
...
+  // Domain is a refinement of the values a metric has. It describes the set
+  // of numbers metric values belong to, if any.
+  enum Domain {
+    // DOMAIN_UNSPECIFIED is the default, and means the metric values do
+    // not belong to any particular domain other than the Type itself.
+    DOMAIN_UNSPECIFIED = 0;
+
+    // NONNEGATIVE is the set of numbers greater than or equal to zero.
+    NONNEGATIVE = 1;
+  }
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:257]: Is this enough to replace monotonic?

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 1 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: I would probably go with just "INVALID"

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 11]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+    //   3. A request is received, the system measures 1 request.
+    //   4. A request is received, the system measures 1 request.
+    //   5. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0 to
+    //      t_0+1 with a value of 3.
+    //   6. A request is received, the system measures 1 request.
+    //   7. A request is received, the system measures 1 request.
+    //   8. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0+1 to
+    //      t_0+2 with a value of 2.
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto]: Kudos for the comment.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 12]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+    //   7. A request is received, the system measures 1 request.
+    //   8. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0 to
+    //      t_0+2 with a value of 5.
+    //   9. The system experiences a fault and loses state.
+    //   10. The system recovers and resumes receiving at time=t_1.
+    //   11. A request is received, the system measures 1 request.
+    //   12. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_1 to
+    //      t_0+1 with a value of 1.
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto]: Another great comment.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 13]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+    CUMULATIVE = 3;
+  }
+
+  // temporality is the Temporality of values this metric has.
+  Temporality temporality = 5;
+
+  // Monotonic is a refinement of the values a metric has. It defines the
+  // relationship values of successively reported metrics have
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto:230]:
```suggestion
  // relationship that values of successively reported metrics have
```

Or maybe

```suggestion
  // relationship between values of successively reported metrics
```

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 1 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: The problem is collision with other enums if they are not prefixed. Both `Type` and `Temporality` would have an `INVALID` and that would cause a collision. The same for `UNSPECIFIED`.

I originally used synonyms, but then there is an unimplemented semantic difference.

Would a suffix work better for you?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 14]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+  // (non-increasing, non-decreasing, or unknown). This is a refinement of
+  // the metric values that can be useful for a receiver in understanding
+  // how to deal with discontinuities in the data (i.e.  calculating
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto:233]:
```suggestion
  // how to deal with discontinuities in the data (i.e. calculating
```

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 4 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:164]: I plan to have comment for time to be kept with the PR to update time: https://github.com/open-telemetry/opentelemetry-proto/pull/137/files#diff-ad849336b0a70e8a18346421908f69edR98-R100

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 5 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto]: changed.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 6 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto]: updated.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 15]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+  // derivates of the data without introducing artifacts from a reset).
+  enum Monotonic {
+    // MONOTONIC_UNSPECIFIED is the default, and means the monotonic nature
+    // of the metric values is unknown.
+    MONOTONIC_UNSPECIFIED = 0;
+
+    // NONDECREASING means all the successive metric values increase or
+    // remain constant.
+    NONDECREASING = 1;
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto:242]: Does `CUMULATIVE` imply `NONDECREASING`? What does a combination of `CUMULATIVE` and `MONOTONIC_UNSPECIFIED` mean? Should `monotonic` field be ignored for `CUMULATIVE`?

Assuming that in the future `Monotonic` enum may be extended with `NONINCREASING` will it mean that a combination of `CUMULATIVE` and `NONINCREASING` is invalid or that the `CUMULATIVE` metric can count downwards?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 16]
```diff
@@ -118,63 +118,150 @@ message MetricDescriptor {
...
+  }
+
+  // monotonic describes the Monotonic refinement of values this metric has.
+  Monotonic monotonic = 6;
+
+  // Domain is a refinement of the values a metric has. It describes the set
+  // of numbers metric values belong to, if any.
```
Comment (tigrannajaryan) [inline opentelemetry/proto/metrics/v1/metrics.proto:249]: I assume this is useful for e.g. visualization purposes, so that one can show a chart with a positive half the axis only? Are there other uses for this field?

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 8 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:234]: What you just describe is a derivative (though I'm guessing you already knew that). I like the more general term derivative, but are you suggesting a preference for rate?

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 9 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:242]: The term increasing/decreasing do not include a value remaining constant.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 10 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:257]: No. The concept of being a positive or zero number is decoupled from the concept of monotonicity.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 15 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:242]: > Does `CUMULATIVE` imply `NONDECREASING`? What does a combination of `CUMULATIVE` and `MONOTONIC_UNSPECIFIED` mean? Should `monotonic` field be ignored for `CUMULATIVE`?
>
> Assuming that in the future `Monotonic` enum may be extended with `NONINCREASING` will it mean that a combination of `CUMULATIVE` and `NONINCREASING` is invalid or that the `CUMULATIVE` metric can count downwards?

While these concepts are commonly associated they are not coupled. For example, if an instrument measures the egress/ingress ratio of network bytes (with ingress being a positive value and egress being a negative value) and sees a total ingress of 1 byte in one cycle and total egress of 2 byte if the instrument reports this as a cumulative count of bytes it would report `1 Byte` and subsequently `-1 Byte`.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 16 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:249]: I imagine there are computational and algorithmic optimizations that can be made as well with this knowledge. But, I think you are correct in identifying visualization concerns as an important use of this information as well.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 4 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:164]: Ehh, on second thought, couldn't hurt.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 1 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: There is no collision because the name of the enum is used as prefix for all the values.

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: I think you'll need to take that up with `protoc` :wink:  :

```
$ make gen-go
...
protoc --go_out=plugins=grpc:/home/tyler/go/src opentelemetry/proto/metrics/v1/metrics.proto
opentelemetry/proto/metrics/v1/metrics.proto:159:5: "INVALID" is already defined in "opentelemetry.proto.metrics.v1.MetricDescriptor".
opentelemetry/proto/metrics/v1/metrics.proto:159:5: Note that enum values use C++ scoping rules, meaning that enum values are siblings of their type, not children of it.  Therefore, "INVALID" must be unique within "opentelemetry.proto.metrics.v1.MetricDescriptor", not just within "Temporality".
make: *** [makefile:25: gen-go] Error 1
```

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: You are correct, I've just saw that in the go generated code, this is ugly because in java this will be:

```java
class MetricDescriptor {
enum Temporality {
  TEMPORALITY_INVALID = 0,
}
}

// And usage
Temporality temporality = Temporality.TEMPORALITY_INVALID
```

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: Another option is to have Temporality defined not a child of MetricDescriptor but as a standalone enum MetricDescriptorTemporality for example. This is just an idea, let me know what you think?

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: Yeah, not great. :cry:

Would a suffix be better?

I like the idea of consistently using the `INVALID` (and `UNSPECIFIED`) term to convey the same meaning (i.e. don't use or there is no defined refinement).

I'm open to suggestions.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 9 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:243]: Hate "english" in Romanian we have "increasing" and "strictly increasing" :))

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:243]: May worth adding a small comment about why not using "increasing"

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 15 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:243]: I think this should be clarify in a comment that it is different than Stackdriver definition. Don't know if we want to point to a specific product but may be good to call that.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 16 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:250]: I am having a hard time understanding the `domain`. If not too much can we remove it for the moment and keep this PR focus only on Temporality and fixing the types?

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 1 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: > Another option is to have Temporality defined not a child of MetricDescriptor but as a standalone enum MetricDescriptorTemporality for example. This is just an idea, let me know what you think?

I was thinking about moving to a different scope, but seems like that would only propagate the problem. Like, why is Temporality in the top scope but not Type? And then it would become less clear when resolving the collision for the Monotonic and Domain defaults of UNSPECIFIED. Which one of those should be in the top scope? What happens if we add another refinement? :shru

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors#metrickind

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: Do you mean suffix just for `INVALID` or for all entries in the enum? I kind of like more `INVALID_TYPE` :)

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:159]: Yeah, just for `INVALID_*` (and `UNSPECIFIED_*`). I'll switch to that.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 9 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:243]: > Hate "english" in Romanian we have "increasing" and "strictly increasing" :))

Oh, let me count the ways ...

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 16 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:250]: > I am having a hard time understanding the `domain`. If not too much can we remove it for the moment and keep this PR focus only on Temporality and fixing the types?

Sure. I'll pull `Domain` and `Monotonic` for another PR.

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto]: Done. @bogdandrutu PTAL.

Comment (tigrannajaryan) [reply opentelemetry/proto/metrics/v1/metrics.proto]: I believe domain was used in the sense of valid values: https://en.wikipedia.org/wiki/Attribute_domain

The term is a bit specialized and may not be familiar to all users of the proto. Perhaps instead use something like "ValueRange" that is more descriptive.

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto]: > I believe domain was used in the sense of valid values: https://en.wikipedia.org/wiki/Attribute_domain
> 
> The term is a bit specialized and may not be familiar to all users of the proto. Perhaps instead use something like "ValueRange" that is more descriptive.

`ValueRange` SGTM. Originally I had posed `Subset` but @yurishkuro pointed out this was an odd name as well: https://github.com/open-telemetry/opentelemetry-proto/pull/137#discussion_r410487868

I'll plan to add it back in as `ValueRange` in the following PR.

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto]: Let's chat about this in the next PR :)

**Comment by bogdandrutu** [comment]
Would be good later to add a table with combination of temporality and types and explain what it is possible what it is not and maybe map to some examples.

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 17]
```diff
@@ -118,63 +118,118 @@ message MetricDescriptor {
...
+    //   8. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0 to
+    //      t_0+2 with a value of 5.
+    //   9. The system experiences a fault and loses state.
+    //   10. The system recovers and resumes receiving at time=t_1.
+    //   11. A request is received, the system measures 1 request.
+    //   12. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_1 to
+    //      t_0+1 with a value of 1.
+    CUMULATIVE = 3;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:224]: One edge case that probably we should document is if the implementation decides to "reset" the `start_time` every export interval making this type essentially a `delta`. What should we recommend/do when we see that case?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 18]
```diff
@@ -118,63 +118,118 @@ message MetricDescriptor {
...
+    //   4. A request is received, the system measures 1 request.
+    //   5. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0 to
+    //      t_0+1 with a value of 3.
+    //   6. A request is received, the system measures 1 request.
+    //   7. A request is received, the system measures 1 request.
+    //   8. The 1 second collection cycle ends. A metric is exported for the
+    //      number of requests received over the interval of time t_0+1 to
+    //      t_0+2 with a value of 2.
+    DELTA = 2;
```
Comment (cijothomas) [inline opentelemetry/proto/metrics/v1/metrics.proto:191]: how is the interval represented when reporting delta? Is it by using `start_time_unix_nano` and `time_unix_nano` fields in `Int64DataPoint` etc, ?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 19]
```diff
@@ -118,63 +118,118 @@ message MetricDescriptor {
...
+  enum Temporality {
+    // INVALID_TEMPORALITY is the default Temporality, it MUST not be
+    // used.
+    INVALID_TEMPORALITY = 0;
+
+    // INSTANTANEOUS is a metric whose values are measured at a particular
+    // instant. The values are not aggregated over any time interval and are
+    // unique per timestamp. As such, these metrics are not expected to have
+    // an associated start time.
+    INSTANTANEOUS = 1;
```
Comment (cijothomas) [inline opentelemetry/proto/metrics/v1/metrics.proto:165]: It'd be great if a clarification is added about `start_time_unix_nano`, `time_unix_nano` in `Int64DataPoint` etc. For INSTANTANEOUS, `start_time_unix_nano` can be null. But what should be the value for `time_unix_nano`?

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 17 above)
Comment (cijothomas) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: This would be the case for Stateless Batcher from Metric Specs.

**Comment by bogdandrutu** [comment]
@cijothomas 

> The part of this proto file which describes Int64DataPoint, DoubleDataPoint, SummaryDataPoint etc. is not updated to reflect the changes in this PR. They still refer to gauge/counter types which are replaced in this PR.

We agree to do that in a separate PR.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 17 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: I think the Stateless Batcher will mark that as a proper `Delta`.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 18 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:191]: That is the plan. Planning to address this issue in a following PR (https://github.com/MrAlias/opentelemetry-proto/commit/5cfae029570935d54ee1b1c903c31a5fb074c44e).

The ask has been to split up https://github.com/open-telemetry/opentelemetry-proto/pull/137 into smaller changes. The assumption that there would be breaking changes or inconsistencies between PRs was something planned and accepted.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 19 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:165]: Same here with regards to this point is expected to be addressed in a future PR (https://github.com/MrAlias/opentelemetry-proto/commit/5cfae029570935d54ee1b1c903c31a5fb074c44e) that incrementally builds to #137.

Comment (cijothomas) [reply opentelemetry/proto/metrics/v1/metrics.proto:165]: Thanks for clarifying! I will look forward to the upcoming PRs.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 17 above)
Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: > One edge case that probably we should document is if the implementation decides to "reset" the `start_time` every export interval making this type essentially a `delta`. What should we recommend/do when we see that case?

I'm not sure how to, nor if we want to, address all the possible ways this can be misused. I think in the situation you describe it would absolutely be an example of incorrectly defining the temporality of the data being sent, and it would not match the documentation already given here, but I don't know how we could handle this other than just processing/resetting every cycle.

Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: Just probably for the moment document clearly that this is a misused of the property.

**Comment by bogdandrutu** [comment]
@cijothomas  are you happy with the responses? Please remove the request changes if so :)

**Comment by cijothomas** [comment]
> @cijothomas are you happy with the responses? Please remove the request changes if so :)

Yes! Just did it.

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 17 above)
Comment (jmacd) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: It sounds like misuse, and it probably will cause poor performance in a downstream system, but the downstream system should be able to handle it correctly according to these definitions. A new start time implies a new cumulative series, so we can see this case as a progression of single-entry series representative deltas of a cumulative instrument.

**Comment by jmacd** [comment]
(I will agree that splitting this work into smaller PRs has helped. Nice work!)

File: opentelemetry/proto/metrics/v1/metrics.proto (hunk 17 above)
Comment (bogdandrutu) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: @MrAlias please maybe just file an issue where you save the current comments, and we can followup. I would prefer not to forget about this discussion.

Comment (MrAlias) [reply opentelemetry/proto/metrics/v1/metrics.proto:224]: @bogdandrutu : created https://github.com/open-telemetry/opentelemetry-proto/issues/143

I'll be sure to address later.

**Comment by bogdandrutu** [comment]
@MrAlias my fault, I thought that will not cause conflicts :(. Can you please rebase?

File: opentelemetry/proto/metrics/v1/metrics.proto [hunk 20]
```diff
@@ -118,59 +118,118 @@ message MetricDescriptor {
...
+    //      t_0+1 with a value of 1.
+    CUMULATIVE = 3;
+  }
+
+  // temporality is the Temporality of values this metric has.
+  Temporality temporality = 5;
+
+  // The set of labels associated with the metric descriptor. Labels in this list apply to
+  // all data points.
+  repeated opentelemetry.proto.common.v1.StringKeyValue labels = 6;
```
Comment (bogdandrutu) [inline opentelemetry/proto/metrics/v1/metrics.proto:232]: This was removed in #144. Please revert this change.

**Comment by davidwitten** [comment]
@MrAlias I'm using this proto to create an OTLP metric exporter in JS. However, I'm confused about how metric kinds map to temporality. I was looking at the [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md), and it didn't mention temporality. Is this correct?

Counter, SumObserver: Cumulative
UpDownCounter,UpDownSumCounter: Delta
ValueRecorder, ValueObserver: Instantaneous

**Comment by jmacd** [comment]
It's not so straightforward, unfortunately. The temporality depends on the aggregation.

For ValueRecorder, ValueObserver: A LastValue or Exact aggregation will have Instaneaneous temporality, whereas a MinMaxLastSumCount, Summary, or Histogram aggregation will have Delta. 

For the other instruments, you have it right:
Counter, SumObserver: Cumulative
UpDownCounter,UpDownSumCounter: Delta

Please keep an eye on https://github.com/open-telemetry/opentelemetry-proto/pull/162

Also see how in https://github.com/open-telemetry/opentelemetry-go/pull/840 an SDK is configured to support converting between cumulative and delta exporters, but when working with OTLP the client should not have to do this. That's why SumObserver is cumulative and Counter is delta in OTLP, it requires no memory in the client--OTLP therefore is considered a PassThrough kind of exporter.

**Comment by davidwitten** [comment]
@jmacd Thanks so much! This was such a big help! I just subscribed to that PR.
